src/corecel/data/detail/LdgIteratorImpl.hh

```diff
@@ -63,6 +63,31 @@ struct LdgLoader<OpaqueId<I, T> const>
 #endif
     }
 };
+
+template<class T>
+struct LdgLoader<T const, std::enable_if_t<std::is_enum_v<T>>>
+{
+    using value_type = T;
+    using pointer = std::add_pointer_t<value_type const>;
+    using reference = value_type;
+    using uderlying_type = std::underlying_type_t<T>;
+
+    CELER_CONSTEXPR_FUNCTION static reference read(pointer p)
+    {
+#if CELER_DEVICE_COMPILE
+        return value_type{__ldg(reinterpret_cast<uderlying_type const*>(p))};
```

Comment:
I am not 100% sure this cast is legal. Looking at the reinterpet_cast rule: 

> Any object pointer type T1* can be converted to another object pointer type cv T2*. This is exactly equivalent to static_cast<cv T2*>(static_cast<cv void*>(expression)) (which implies that if T2's alignment requirement is not stricter than T1's, the value of the pointer does not change and conversion of the resulting pointer back to its original type yields the original value). In any case, the resulting pointer may only be dereferenced safely if allowed by the type aliasing rules (see below). 

The enum and its underlying type have the same size and alignment requirements so casting is fine but I think that dereferencing, which is probably done in `__ldg` implementation, breaks the aliasing rules.

Reply:
In practice, it _most likely_ doesn't matter since we don't keep around the pointer obtained from `reinterpret_cast` there shouldn't be an opportunity for the compiler to take advantage of this rule. 

Still, if we want to be safe, [bit_cast](https://en.cppreference.com/w/cpp/numeric/bit_cast) would be useful, maybe we can use the suggested implementation using [std::memcpy](https://en.cppreference.com/w/cpp/string/byte/memcpy) to get around it?

Reply:
Yeah, I'd guess you're right. Oh well! Given the little we've gained yet, I suspect we won't see much by supporting enums. If we really wanted to though, we could store them on device as their underlying type, and static cast to return them? Kind of like how `range` works for enum iterators.

Reply:
Yeah, I guess we could use memcpy to avoid reinterpreting on the return type, but we have to reinterpret_cast the pointer to give it to `__ldg` right?

Reply:
If you look at the PTX output of [this example](https://godbolt.org/z/qYezvhvhx), the first two blocks are exactly the same, so nvcc can optimize the memcpy to a cast. And yeah, that's pretty much for nothing supporting enum will most likely not make a difference...

Reply:
Not sure if that's the best example since it's doing everything locally? Here's an example of [loading enum directly versus using ldg](https://godbolt.org/z/jsbP793sW)... looks like casting or not ends up the same. (Using ldg or not changes the load command from `u64` to `s64`,  not sure if that matters...)

Reply:
mmh yeah, I've added [one example](https://godbolt.org/z/bee5q7j8n) with the memcpy and static_cast, it is doing quite a bit more than just the reinterpret_cast. I suppose the main difference in the ldg load command is `.nc` for the non-coherent cache, not sure why it's switching between signed/unsigned. So, since `cast2` and `cast3` are the same, should we just go with the `reinterpret_cast`?

Reply:
Yes. Let's either reinterpret_cast (only needed for `__ldg(value)` on device) or elect not to try to accelerate enums.

Reply:
Alright, so with that I think we're good!

Reply:
Kinda surprising that nvcc does the actual copy whereas gcc/clang [optimize it](https://godbolt.org/z/voqh373E7) to a single move.